### ISSUE TITLE
Make dedupe more efficient

### ISSUE THREAD

**Issue by ncw** [rclone/rclone#1700]
At the moment dedupe builds a map of the entire tree before doing anything which

  * uses up lots of memory
  * makes it do nothing for ages

Fix this by deduping each directory at a time.

**Comment by timofonic** [comment]
Please ignore my message, I have no idea. But maybe other project developed in Go solved it...

| Project                             | Description                                     |         License        | First commit | Last commit  |
|-------------------------------------|-------------------------------------------------|:----------------------:|--------------|--------------|
| [restic](/restic/restic)            | Fast, secure, efficient backup program          |      BSD 2-Clause      | Apr 6, 2014  | Sep 26, 2017 |
| [kopia](/kopia/kopia)               | Simple Personal Cloud Backup                    |       Apache-2.0       | Dec 19, 2015 | Sep 17, 2017 |
| [walk](/whosonfirst/walk)           | Fast parallel version of golang filepath.Walk() |      BSD 3-Clause      | Jul 6, 2014  | Aug 3, 2016  |
| [duplicacy](/gilbertchen/duplicacy) | A new generation cloud backup tool              | non-commercial  custom | Feb 23, 2016 | Sep 26, 2017 |

**Comment by mightyfree** [comment]
I move significant amounts of data TB+ into Google Drive using rclone, often with folders of hundreds of large images. Having dedupe is great, but merging these files and folders take forever! Perhaps the delete operation can take place before the merging one - is this possible? I'm all for more memory use to map everything, assuming the merging time is dramatically decreased. Thoughts?

**Comment by ncw** [comment]
You are right the merge operation is less efficient than it could be. I wrote it for correctness - it is now time to examine the efficiency!

The merge take much longer than it needs to for deeply nested directories so when the directories being merged have lots of subdirectories. Is that the case you are coming across?

Deduping directory at a time will help with this a lot I think.

**Comment by egzakharovich** [comment]
Postponing deletions and doing them in bulk can also be helpful. Have a `rclone dedupe --by-hash --dedupe-mode first` command running right now on a thousands of small files. And it's going through the night like:

> match
> delete
> match
> delete
> match
> delete
>....
